src/game/audio.rs

```diff
@@ -0,0 +1,154 @@
+
+pub struct AudioAssets {
+    bg: Handle<AudioSource>,
+    menu: Handle<AudioSource>,
+    hit1: Handle<AudioSource>,
+    hit2: Handle<AudioSource>,
+    hit3: Handle<AudioSource>,
+    death: Handle<AudioSource>,
+    drink1: Handle<AudioSource>,
+    drink2: Handle<AudioSource>,
+    drink3: Handle<AudioSource>,
+    shoot: Handle<AudioSource>,
+    fast_shoot: Handle<AudioSource>,
+    lvlup: Handle<AudioSource>,
+    bg_channel: AudioChannel,
+    menu_channel: AudioChannel,
+}
+
+impl Plugin for GameAudioPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugin(AudioPlugin)
+            .add_startup_system_to_stage(StartupStage::PreStartup, load_audio)
+            .add_system_set(SystemSet::on_enter(AppState::InGame).with_system(play_bg_music))
+            .add_system_set(SystemSet::on_exit(AppState::InGame).with_system(play_menu_music))
+            .add_system(play_hit_sfx)
+            .add_system(play_death_sfx)
+            .add_system(play_eat_sfx)
+            .add_system(play_lvlup_sfx)
+            .add_system(play_shoot_sfx)
+            .add_system(play_fast_shoot_sfx)
+            .add_startup_system(play_menu_music)
+            .add_event::<AudioRustEvent>()
+            .add_event::<AudioCoffeeEvent>()
+            .add_event::<AudioHitEvent>()
+            .add_event::<AudioFastShootEvent>()
+            .add_event::<AudioShootEvent>()
+            .add_event::<AudioDeadPlayerEvent>();
+    }
+}
+
+pub fn play_hit_sfx(
+    audio: Res<Audio>,
+    audio_state: Res<AudioAssets>,
+    mut audio_event: EventReader<AudioHitEvent>,
+) {
+    audio_event.iter().for_each(|_| {
+        let mut rng = thread_rng();
+        match rng.gen_range(0..100) {
+            // 1/3 chance of playing each sound
+            0..=33 => audio.play(audio_state.hit1.clone()),
+            34..=67 => audio.play(audio_state.hit2.clone()),
+            _ => audio.play(audio_state.hit3.clone()),
+        };
+    });
+}
+
+pub fn play_eat_sfx(
+    audio: Res<Audio>,
+    audio_state: Res<AudioAssets>,
+    mut audio_event: EventReader<AudioCoffeeEvent>,
+) {
+    audio_event.iter().for_each(|_| {
+        let mut rng = thread_rng();
+        match rng.gen_range(0..100) {
+            // 1/3 chance of playing each sound
+            0..=33 => audio.play(audio_state.drink1.clone()),
+            34..=67 => audio.play(audio_state.drink2.clone()),
+            _ => audio.play(audio_state.drink3.clone()),
+        };
+    });
+}
+
+pub fn play_lvlup_sfx(
+    audio: Res<Audio>,
+    audio_state: Res<AudioAssets>,
+    mut audio_event: EventReader<AudioRustEvent>,
+) {
+    audio_event.iter().for_each(|_| {
+        audio.play(audio_state.lvlup.clone());
+    });
+}
+
+pub fn play_shoot_sfx(
+    audio: Res<Audio>,
+    audio_state: Res<AudioAssets>,
+    mut audio_event: EventReader<AudioShootEvent>,
+) {
+    audio_event.iter().for_each(|_| {
+        audio.play(audio_state.shoot.clone());
+    });
+}
+
+pub fn play_fast_shoot_sfx(
```

Comment:
hm, no powtarza się ten kod, ale nie wiem, czy bez makr dałoby się to jakoś lepiej zrobić...

Reply:
o, wiem! może event mógłby mieć po prostu parametr oznaczający rodzaj dźwięku - i wtedy wszystko w jednym systemie dałoby się ogarnąć